### PR TITLE
Add --chown flag to Dockerfile ADD and COPY

### PR DESCRIPTION
**Pull request by duglin**
Carrying moby/moby#27303 since Kara is off doing other exciting things.

- What Kara did
Add a --chown flag to Dockerfile ADD and COPY instructions, proposed by moby/moby#13600

- How Kara did it
Added flag to add and disptachCopy of builder/dockerfile/dispatchers.go, and passed the relevant information through to the daemon which looks up the user/group using a new function on the container object.

- How to verify it
Create a Dockerfile that uses the --chown flag on an ADD or COPY instruction, and check that the file is listed as having the appropriate owner

- Description for the changelog
add --chown flag to Dockerfile ADD and COPY instructions

2nd commit is just a rebase.

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

File: daemon/oci_windows.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ func (daemon *Daemon) createSpec(c *container.Container) (*specs.Spec, error) {
 	}
 
 	// TODO Windows - this can be removed. Not used (UID/GID)
-	rootUID, rootGID := daemon.GetRemappedUIDGID()
```
Comment (jhowardmsft) [inline daemon/oci_windows.go]: See https://github.com/docker/docker/pull/28594#issuecomment-261604763.

**Comment by thaJeztah** [comment]
Let's move this to code review

**Comment by tianon** [comment]
If `--chmod` might come later, would it make sense to call this `--chown` instead of `--user` to make it more clear what it's doing?

**Comment by duglin** [comment]
did someone suggest a chmod option? I missed that

**Comment by tianon** [comment]
It's been discussed a few times for similar reasons to wanting to control
the user, and there's a comment on this PR asking about it specifically.
:smile:

**Comment by duglin** [comment]
doi - sorry I missed that. I was scanning too quickly.

**Comment by jrusiecki** [comment]
I need ADD extension  to use instead of:
`run wget ... | chmod +x ... `
  which is superior (layer-wise) to 
```
ADD ...
run chmod +x ...
```
Thus without **+x** option ADD is useless anyway (if I already have run... I put there also non-executables so will not use COPY/ADD even for --user option).
COPY would benefit from +x too (although file copied from linux is probably already +x) as it would improve quality by clearly stating +x so to limit errors.

**Comment by Asher256** [comment]
+1

**Comment by duglin** [comment]
@tianon which would be better:
`--user` && `--mode`
or
`--chown` && `--chmod` 
?

The 2nd feels a little like we're trying to expose our implementation detail (changing things after the fact vs setting the properties on the files during creation) - when the end user should just care about the properties having the right values regardless of how they are set.

**Comment by tianon** [comment]
Hmm, good point, but `--user` is already in use for changing the runtime user, so potentially confusing double usage.  For what it's worth, `rsync` uses `--chown` and `--chmod`:
```
     --chown=USER:GROUP      simple username/groupname mapping
     --chmod=CHMOD           affect file and/or directory permissions
```

In my mind, the argument for `--chown` and `--chmod` is that normally, `ADD`/`COPY` try as hard as possible to preserve the _source_ ownership/permissions, so the "change" implied by `--chmod` and `--chown` as the name is referring to the change which has to happen during the copy (ie, "copy XYZ, but change the owner to ABC").

**Comment by glensc** [comment]
i'm also proposing `DEFATTR` so that if you use multiple `COPY` or `ADD` you don't have to repeat the options:

```
DEFATTR chown=USER
DEFATTR chown=USER:GROUP
DEFATTR chown=UID
DEFATTR chown=UID:GID
```

`chmod` is a bit tricky because you definetely want different permissions for files and dirs:
```
DEFATTR chmod=a+rX
DEFATTR chmod=a+X
```

these would rise the `x` bit only if `x` bit is present for any of the users: `chmod(1)`: 
> execute/search  only if the file is a directory or already has execute permission for some user (X)

or more clearer directives for different kinds:
```
DEFATTR dir_chmod=0755 file_chmod=0644
```

**Comment by thaJeztah** [comment]
@glensc could you open a separate issue for that proposal? Also, would be worth describing how it should interact with different commands that take options (just for illustration, if `LABEL` would take `--chown` as an option).

**Comment by duglin** [comment]
@tianon ok - will go with `--chown`.

I'd prefer to leave `--chmod` for another PR because I think more thinking needs to go into that.

I have a high-order question for people.... if we were developing `docker build` from scratch would we have a `--chown` flag at all on ADD/COPY or would we have `USER` impact subsequent `RUN`, `ADD` and `COPY` commands?  I'm asking because @glensc's latest comment got me wondering if normally people would have expected `USER` to have a broader impact than it does today.  I'm not suggesting we break backwards compatibility, but I am wondering if instead of a flag on `ADD`/`COPY` we should consider a flag on `USER` instead - something like: `USER --chown john:mygroup`. This would mean use `john:mygroup` for all subsequent `RUN` commands (as well as the CMD) like it does today, but also use `john:mygroup` on all subsequent ADD/COPY commands too - which is probably what most newbies would have expected anyway.

To be clear, the `--chown` flag doesn't take any args, the `john:mygroup` is the normal USER arg.  Perhaps `--chown` isn't the best word since it might be mentally parsed incorrectly, so perhaps `--copy` instead - dunno, I'm open to other words.

thoughts?

**Comment by glensc** [comment]
i definitely wouldn't tie `USER` with `COPY`/`ADD`. `USER` changes effective uid of running process (in build and later in run), `COPY`/`ADD` i want always add root owned files so `USER` would not able to modify them.

**Comment by glensc** [comment]
@thaJeztah: #29853 altho i'm not sure what you mean `LABEL` taking `chown`, it's just metadata not related to files.

**Comment by thaJeztah** [comment]
@glensc the `LABEL` example does not make sense, I know, but just to illustrate that "other" commands may possibly have a `--chown` option, and I _assume_ defaults should be set per command, not globally (by default)

**Comment by duglin** [comment]
modified flag to be `--chown`

**Comment by jrusiecki** [comment]
I've copied my comments to #29853.

**Comment by jazoom** [comment]
This still isn't in the docs. Is it used like this?

COPY from.txt /to.txt --chown someuser

**Comment by cpuguy83** [comment]
@jazoom This is not merged.

**Comment by dmytroleonenko** [comment]
Looking for the solution for ages. Subscribed

File: container/container.go [hunk 2]
```diff
@@ -300,6 +301,36 @@ func (container *Container) GetRootResourcePath(path string) (string, error) {
 	return symlink.FollowSymlinkInScope(filepath.Join(container.Root, cleanPath), container.Root)
 }
 
+// ParseUserGrp takes `username` in the format of username, uid, username:groupname,
+// uid:gid, username:gid, or uid:groupname and parses the passwd file in the container
+// to return the ExecUser referred to by `username`.
+func (container *Container) ParseUserGrp(username string) (*user.ExecUser, error) {
+	passwdPath, err := user.GetPasswdPath()
```
Comment (dnephin) [inline container/container.go]: I believe this call always returns an err on windows, which means this new flag would not work for the windows builder.

Are we ok with that?

**Comment by dnephin** [comment]
Needs a rebase as well

**Comment by dnephin** [comment]
@duglin Thanks for working on this!

It seems this PR hasn't seen any activity in 4 weeks. I'm going to close the PR since there is no activity. If/when you have time to rebase, and respond to the comments please feel free to re-open it (or ping me and I can).

**Comment by duglin** [comment]
@dnephin this hasn't been touched in a while because people asked for the discussion to be moved over to:  https://github.com/docker/docker/issues/30110

I can rebase now if we want but we really need #30110 to be done/discussed first.

**Comment by dnephin** [comment]
Ok, let's reopen once the discussion in #30110 sorted

**Comment by italomaia** [comment]
I'm unsure how #32816 is blocking this issue. To some extent, they seem unrelated as the problem of 32816 is related to the source, right? This PR seems so helpful it hurts to see it goes by closed for another week.

**Comment by thaJeztah** [comment]
reopening (again) following https://github.com/moby/moby/issues/30110#issuecomment-316790885, but probably needs a bit of rewriting

**Comment by thaJeztah** [comment]
(make that a _lot_ of rewriting LOL)

**Comment by estesp** [comment]
I can carry this if that's helpful to @duglin?

**Comment by duglin** [comment]
go for it! thanks @estesp

**Comment by estesp** [comment]
Carried in #34263

**Pull request closed without merging**